File: src/utils/irpfRetencion.ts
/** AEAT-style IRPF withholding (Modelo 145 inputs) — mirror of backend/app/irpf_retencion.py */

export type FamilySituation = "1" | "2" | "3";
export type DisabilityDegree = "none" | "33_64" | "65_plus";
export type ContractType = "indefinido" | "temporal" | "especial";

export type IrpfDependent = {
  kind: "descendant" | "ascendant";
  age: number;
  disability: DisabilityDegree;
  shared_custody?: boolean;
  mobility_reduced?: boolean;
};

export type IrpfRetencionInput = {
  annual_gross: number;
  age: number;
  family_situation?: FamilySituation;
  disability?: DisabilityDegree;
  mobility_reduced?: boolean;
  geographic_mobility?: boolean;
  dependents?: IrpfDependent[];
  contract_type?: ContractType;
  pagas?: number;
  ss_pct?: number;
};

export type IrpfRetencionResult = {
  irpf_pct: number;
  ss_pct: number;
  neto_estimado: number;
  irpf_amount: number;
  ss_amount: number;
  annual_irpf: number;
  family_minimum: number;
  exclusion_limit: number;
  retention_base: number;
  disclaimer: string;
};

function truncate2(value: number): number {
  return Math.floor(value * 100 + 1e-9) / 100;
}

function annualRetentionQuota(base: number): number {
  const brackets: [number, number][] = [
    [12_450, 0.19],
    [20_200, 0.24],
    [35_200, 0.3],
    [60_000, 0.37],
    [300_000, 0.45],
    [Infinity, 0.47],
  ];
  let remaining = Math.max(base, 0);
  let prev = 0;
  let tax = 0;
  for (const [limit, rate] of brackets) {
    const width = Math.min(Math.max(limit - prev, 0), remaining);
    if (width <= 0) {
      prev = limit;
      continue;
    }
    tax += width * rate;
    remaining -= width;
    prev = limit;
    if (remaining <= 0) break;
  }
  return Math.max(tax, 0);
}

function art20(rendNeto: number): number {
  if (rendNeto <= 14_747.5) return 7_302;
  if (rendNeto <= 19_747.5) return Math.max(0, 7_302 - 1.14 * (rendNeto - 14_747.5));
  return 0;
}

function disabilityMin(degree: DisabilityDegree, mobility = false): number {
  if (degree === "65_plus" || mobility) return 9_000;
  if (degree === "33_64") return 3_000;
  return 0;
}

function personalMin(age: number, disability: DisabilityDegree, mobility = false): number {
  let base = 5_550;
  if (age >= 75) base += 1_150 + 1_400;
  else if (age >= 65) base += 1_150;
  return base + disabilityMin(disability, mobility);
}

function descendantsMin(deps: IrpfDependent[]): number {
  const kids = deps.filter((d) => d.kind === "descendant").sort((a, b) => a.age - b.age);
  const ordinal = [2_400, 2_700, 4_000];
  let total = 0;
  kids.forEach((kid, i) => {
    let amount = i < 3 ? ordinal[i]! : 4_500;
    if (kid.age < 3) amount += 2_800;
    amount += disabilityMin(kid.disability, kid.mobility_reduced);
    if (kid.shared_custody) amount *= 0.5;
    total += amount;
  });
  return total;
}

function ascendantsMin(deps: IrpfDependent[]): number {
  let total = 0;
  for (const dep of deps) {
    if (dep.kind !== "ascendant") continue;
    let amount = 1_150;
    if (dep.age >= 75) amount += 1_400;
    amount += disabilityMin(dep.disability, dep.mobility_reduced);
    total += amount;
  }
  return total;
}

function exclusionLimit(situation: FamilySituation, nDesc: number): number {
  const base = { "1": 17_197, "2": 17_197, "3": 15_876 }[situation];
  const extra = { "1": 2_800, "2": 2_100, "3": 1_800 }[situation];
  return base + Math.max(nDesc, 0) * extra;
}

export function calculateIrpfRetencion(input: IrpfRetencionInput): IrpfRetencionResult {
  const annualGross = Math.max(Number(input.annual_gross) || 0, 0);
  const pagas = Math.max(Number(input.pagas) || 14, 12);
  const age = Math.max(Number(input.age) || 0, 0);
  const situation = input.family_situation ?? "3";
  const disability = input.disability ?? "none";
  const mobility = Boolean(input.mobility_reduced);
  const geo = Boolean(input.geographic_mobility);
  const contract = input.contract_type ?? "indefinido";
  const dependents = input.dependents ?? [];
  const nDesc = dependents.filter((d) => d.kind === "descendant").length;

  let ssPct = input.ss_pct;
  if (ssPct == null) ssPct = contract === "temporal" ? 6.55 : 6.5;
  const ssRate = Math.max(ssPct, 0) / 100;
  const annualSs = annualGross * ssRate;
  const rendNeto = Math.max(annualGross - annualSs, 0);

  let otherExpenses = 2_000;
  if (disability !== "none" || mobility) {
    otherExpenses += disability === "65_plus" || mobility ? 3_500 : 2_000;
  }
  if (geo) otherExpenses += 2_000;

  const retentionBase = Math.max(rendNeto - otherExpenses - art20(rendNeto), 0);
  const spouseMin = situation === "2" ? 3_400 : 0;
  const familyMin =
    personalMin(age, disability, mobility) +
    descendantsMin(dependents) +
    ascendantsMin(dependents) +
    spouseMin;

  let raw = Math.max(annualRetentionQuota(retentionBase) - annualRetentionQuota(familyMin), 0);
  const excl = exclusionLimit(situation, nDesc);
  if (annualGross <= excl) raw = 0;
  else if (annualGross <= 35_200) {
    raw = Math.min(raw, Math.max(annualGross - 15_876, 0) * 0.43);
  }

  let irpfPct = annualGross > 0 && raw > 0 ? truncate2((raw / annualGross) * 100) : 0;
  if (contract === "temporal" && irpfPct > 0) irpfPct = Math.max(irpfPct, 2);
  if (contract === "especial") irpfPct = Math.max(irpfPct, 15);

  const brutoMensual = annualGross / pagas;
  const irpfAmount = brutoMensual * irpfPct / 100;
  const ssAmount = brutoMensual * ssRate;
  const neto = Math.max(brutoMensual - irpfAmount - ssAmount, 0);

  return {
    irpf_pct: irpfPct,
    ss_pct: Math.round(ssRate * 10000) / 100,
    neto_estimado: Math.round(neto * 100) / 100,
    irpf_amount: Math.round(irpfAmount * 100) / 100,
    ss_amount: Math.round(ssAmount * 100) / 100,
    annual_irpf: Math.round(annualGross * irpfPct) / 100,
    family_minimum: Math.round(familyMin * 100) / 100,
    exclusion_limit: Math.round(excl * 100) / 100,
    retention_base: Math.round(retentionBase * 100) / 100,
    disclaimer:
      "Estimación orientativa según algoritmo AEAT 2026 y datos tipo Modelo 145. No sustituye el Servicio de Cálculo de Retenciones oficial ni la nómina del pagador.",
  };
}

File: backend/app/irpf_retencion.py
"""AEAT-style IRPF withholding calculator (rendimientos del trabajo).

Implements the core of the Agencia Tributaria retention algorithm for 2026,
driven by the same personal/family inputs as Modelo 145 (comunicación de
datos al pagador). Orientative — not a substitute for the official AEAT
service or payroll software.
"""
from __future__ import annotations

from dataclasses import dataclass
from math import floor
from typing import Any, Dict, List, Literal, Optional


FamilySituation = Literal["1", "2", "3"]
DisabilityDegree = Literal["none", "33_64", "65_plus"]
ContractType = Literal["indefinido", "temporal", "especial"]


@dataclass
class Dependent:
    """Descendant or ascendant for mínimo familiar."""

    kind: Literal["descendant", "ascendant"]
    age: int
    disability: DisabilityDegree = "none"
    shared_custody: bool = False  # 50% of descendant minimum
    mobility_reduced: bool = False


def truncate2(value: float) -> float:
    """AEAT truncates the withholding rate to 2 decimal places (not round)."""
    return floor(value * 100.0 + 1e-9) / 100.0


def annual_retention_quota(base: float) -> float:
    """Progressive retention scale AEAT 2026 (absolute limits)."""
    brackets = [
        (12_450.0, 0.19),
        (20_200.0, 0.24),
        (35_200.0, 0.30),
        (60_000.0, 0.37),
        (300_000.0, 0.45),
        (float("inf"), 0.47),
    ]
    remaining = max(base, 0.0)
    prev = 0.0
    tax = 0.0
    for limit, rate in brackets:
        width = min(max(limit - prev, 0.0), remaining)
        if width <= 0:
            prev = limit
            continue
        tax += width * rate
        remaining -= width
        prev = limit
        if remaining <= 0:
            break
    return max(tax, 0.0)


def art20_trabajo_reduction(rend_neto: float) -> float:
    """Reducción art. 20 LIRPF (estimación 2026)."""
    # Intervalos alineados con la actualización 2026 (máx. 7.302 €).
    if rend_neto <= 14_747.5:
        return 7_302.0
    if rend_neto <= 19_747.5:
        return max(0.0, 7_302.0 - 1.14 * (rend_neto - 14_747.5))
    return 0.0


def disability_minimum(degree: DisabilityDegree, mobility_reduced: bool = False) -> float:
    if degree == "65_plus" or mobility_reduced:
        return 9_000.0
    if degree == "33_64":
        return 3_000.0
    return 0.0


def personal_minimum(age: int, disability: DisabilityDegree, mobility_reduced: bool = False) -> float:
    base = 5_550.0
    if age >= 75:
        base += 1_150.0 + 1_400.0
    elif age >= 65:
        base += 1_150.0
    base += disability_minimum(disability, mobility_reduced)
    return base


def descendants_minimum(dependents: List[Dependent]) -> float:
    kids = [d for d in dependents if d.kind == "descendant"]
    # Order: youngest first for under-3 bonus, ordinal amounts by declaration order.
    kids_sorted = sorted(kids, key=lambda d: d.age)
    total = 0.0
    ordinal_amounts = [2_400.0, 2_700.0, 4_000.0]  # 4th+ → 4_500
    for i, kid in enumerate(kids_sorted):
        amount = ordinal_amounts[i] if i < 3 else 4_500.0
        if kid.age < 3:
            amount += 2_800.0
        amount += disability_minimum(kid.disability, kid.mobility_reduced)
        if kid.shared_custody:
            amount *= 0.5
        total += amount
    return total


def ascendants_minimum(dependents: List[Dependent]) -> float:
    total = 0.0
    for dep in dependents:
        if dep.kind != "ascendant":
            continue
        amount = 1_150.0
        if dep.age >= 75:
            amount += 1_400.0
        amount += disability_minimum(dep.disability, dep.mobility_reduced)
        total += amount
    return total


def exclusion_limit(situation: FamilySituation, n_descendants: int) -> float:
    """Annual gross under which withholding is zero (simplified AEAT table 2026)."""
    # Base rows without descendants (approx. AEAT informative table 2026).
    base = {"1": 17_197.0, "2": 17_197.0, "3": 15_876.0}[situation]
    # Extra per descendant — coarse approximation of the AEAT table steps.
    extra_per_child = {"1": 2_800.0, "2": 2_100.0, "3": 1_800.0}[situation]
    return base + max(n_descendants, 0) * extra_per_child


def apply_43pct_cap(annual_gross: float, raw_retention: float) -> float:
    """For gross ≤ 35_200, retention cannot exceed 43% of excess over 15_876-ish band."""
    if annual_gross <= 0 or annual_gross > 35_200:
        return raw_retention
    # Simplified cap used by AEAT around the general exclusion threshold.
    excess = max(annual_gross - 15_876.0, 0.0)
    cap = excess * 0.43
    return min(raw_retention, cap)


def ss_rate_for_contract(contract_type: ContractType) -> float:
    # CC 4.70 + desempleo 1.55/1.60 + FP 0.10 + MEI 0.15
    if contract_type == "temporal":
        return 6.55
    return 6.50


def calculate_retencion(
    *,
    annual_gross: float,
    age: int,
    family_situation: FamilySituation = "3",
    disability: DisabilityDegree = "none",
    mobility_reduced: bool = False,
    dependents: Optional[List[Dependent]] = None,
    contract_type: ContractType = "indefinido",
    pagas: int = 14,
    ss_pct: Optional[float] = None,
    geographic_mobility: bool = False,
) -> Dict[str, Any]:
    annual_gross = max(float(annual_gross), 0.0)
    pagas = max(int(pagas), 12)
    dependents = dependents or []
    n_desc = sum(1 for d in dependents if d.kind == "descendant")

    if ss_pct is None:
        ss_pct = ss_rate_for_contract(contract_type)
    ss_rate = max(float(ss_pct), 0.0) / 100.0
    annual_ss = annual_gross * ss_rate

    rend_neto = max(annual_gross - annual_ss, 0.0)
    other_expenses = 2_000.0
    if disability in ("33_64", "65_plus") or mobility_reduced:
        # Incremento gastos por discapacidad del trabajador (simplificado).
        other_expenses += 3_500.0 if disability == "65_plus" or mobility_reduced else 2_000.0
    if geographic_mobility:
        other_expenses += 2_000.0

    art20 = art20_trabajo_reduction(rend_neto)
    retention_base = max(rend_neto - other_expenses - art20, 0.0)

    min_personal = personal_minimum(age, disability, mobility_reduced)
    min_desc = descendants_minimum(dependents)
    min_asc = ascendants_minimum(dependents)
    # Situación 2: cónyuge sin rentas → mínimo familiar adicional del cónyuge (simplificado).
    spouse_min = 3_400.0 if family_situation == "2" else 0.0
    family_min = min_personal + min_desc + min_asc + spouse_min

    quota = annual_retention_quota(retention_base)
    quota_min = annual_retention_quota(family_min)
    raw_retention = max(quota - quota_min, 0.0)

    excl = exclusion_limit(family_situation, n_desc)
    if annual_gross <= excl:
        raw_retention = 0.0
    else:
        raw_retention = apply_43pct_cap(annual_gross, raw_retention)

    if annual_gross > 0 and raw_retention > 0:
        irpf_pct = truncate2((raw_retention / annual_gross) * 100.0)
    else:
        irpf_pct = 0.0

    # Minimum rates by contract type when retention would otherwise be positive / zero rules.
    if contract_type == "temporal" and irpf_pct > 0:
        irpf_pct = max(irpf_pct, 2.0)
    elif contract_type == "especial":
        irpf_pct = max(irpf_pct, 15.0)

    annual_irpf = annual_gross * irpf_pct / 100.0
    bruto_mensual = annual_gross / pagas
    irpf_amount = bruto_mensual * irpf_pct / 100.0
    ss_amount = bruto_mensual * ss_rate
    neto = max(bruto_mensual - irpf_amount - ss_amount, 0.0)

    return {
        "annual_gross": round(annual_gross, 2),
        "pagas": pagas,
        "bruto_mensual": round(bruto_mensual, 2),
        "contract_type": contract_type,
        "family_situation": family_situation,
        "age": int(age),
        "ss_pct": round(ss_rate * 100.0, 2),
        "irpf_pct": irpf_pct,
        "ss_amount": round(ss_amount, 2),
        "irpf_amount": round(irpf_amount, 2),
        "neto_estimado": round(neto, 2),
        "annual_ss": round(annual_ss, 2),
        "annual_irpf": round(annual_irpf, 2),
        "rend_neto": round(rend_neto, 2),
        "other_expenses": round(other_expenses, 2),
        "art20_reduction": round(art20, 2),
        "retention_base": round(retention_base, 2),
        "family_minimum": round(family_min, 2),
        "exclusion_limit": round(excl, 2),
        "quota": round(quota, 2),
        "quota_minimum": round(quota_min, 2),
        "n_descendants": n_desc,
        "source": "modelo145-aeat2026",
        "disclaimer": (
            "Estimación orientativa según algoritmo AEAT 2026 y datos tipo Modelo 145. "
            "No sustituye el Servicio de Cálculo de Retenciones oficial ni la nómina del pagador."
        ),
    }


def dependents_from_payload(raw: Optional[List[Dict[str, Any]]]) -> List[Dependent]:
    out: List[Dependent] = []
    for item in raw or []:
        if not isinstance(item, dict):
            continue
        kind = item.get("kind") or "descendant"
        if kind not in ("descendant", "ascendant"):
            continue
        degree = item.get("disability") or "none"
        if degree not in ("none", "33_64", "65_plus"):
            degree = "none"
        try:
            age = int(item.get("age", 0))
        except (TypeError, ValueError):
            age = 0
        out.append(
            Dependent(
                kind=kind,  # type: ignore[arg-type]
                age=max(age, 0),
                disability=degree,  # type: ignore[arg-type]
                shared_custody=bool(item.get("shared_custody")),
                mobility_reduced=bool(item.get("mobility_reduced")),
            )
        )
    return out
